teps/0002-custom-tasks.md

```diff
@@ -0,0 +1,392 @@
+```
+
+This references an `Example` CRD type defined by the custom task author, an instance of which is named `my-example`.
+
+When a `Run` object is created, Tekton will validate that the `ref` is specified, and that the specified CRD type is defined, using webhook validation.
+
+After that, Tekton Pipelines expects a custom task author to implement a controller for `Run` objects that reference their type (annotated throughout this proposal with the shorthand `Run<Example>`) to take some action, and eventually update its `.status` to signal completion, either successfully or unsuccessfully, using the `conditions` model used by Tekton PipelineRuns and TaskRuns.
+
+### Validation
+
+Custom Task authors can implement webhook validation for CR objects of their provided type (e.g., to validate `Example` object definitions). Validation is optional but recommended, and examples and sample frameworks will demonstrate this functionality.
+
+### Status Reporting
+
+When the `Run<Example>` is validated and created, the Custom Task controller should be notified and begin doing some operation. When the operation begins, the controller should update the `Run`'s `.status.conditions` to report that it's ongoing:
+
+```
+status
+  conditions:
+  - type: Succeeded
+    status: Unknown
+```
+
+When the operation completes, if it was successful, the condition should report `status: True`, and optionally a brief `reason` and human-readable `message`:
+
+```
+status
+  conditions:
+  - type: Succeeded
+    status: True
+    reason: ExampleComplete
+    message: Yay, good times
+```
+
+If the operation was _unsuccessful_, the condition can report `status: False`, and optionally a `reason` and human-readable `message`:
+
+```
+status
+  conditions:
+  - type: Succeeded
+    status: False
+    reason: ExampleFailed
+    message: Oh no bad times
+```
+
+The `Run` type's `.status` will also allow controllers to report other fields, such as `startTime`, `completionTime`, `results` (see below), and arbitrary context-dependent fields the Custom Task author wants to report. A fully-specified `Run` status might look like:
+
+```
+status
+  conditions:
+  - type: Succeeded
+    status: True
+    reason: ExampleComplete
+    message: Yay, good times
+  completionTime: "2020-06-18T11:55:01Z"
+  startTime: "2020-06-18T11:55:01Z"
+  results:
+  - name: first-name
+    value: Bob
+  - name: last-name
+    value: Smith
+  arbitraryField: hello world
+  arbitraryStructuredField:
+    listOfThings: ["a", "b", "c"] 
+```
+
+### Pipeline Integration
+
+Enabling `Run`s by themselves are not terribly compelling. Their power comes from being specified in `Pipeline`s, and executed during `PipelineRun`s.
+
+Under this proposal, a user can define a Pipeline that invokes a Custom Task, specified similar to how `Task`s are specified today:
+
+```
+apiVersion: tekton.dev/v1beta1
+kind: Pipeline
+metadata:
+  name: example-pipeline
+spec:
+  tasks:
+  …
+  - name: example-task
+    taskRef:
```

Comment:
Meh, I kind of like the simplicity of reusing the existing field, though it does end up being the only place in the API we actually refer to custom runnable objects as "tasks". 🤔

Reply:
It is also a non-flag path to `Run<Task,ClusterTask,Pipeline>` vs. `{Task,ClusterTask,Pipline}Run` 😇

Reply:
We also *may* want to transition from `taskRef` to `ref` (independently of this proposal) — we know we are on a task, so `taskRef` is redondant, `ref` would be enough. In any case, this would be another proposal :wink: